ramanujan/enumerators/FREnumerator.py

```diff
@@ -0,0 +1,123 @@
+            if num_of_calculated_vals >= 3 and \
+                    abs(calculated_values[-2] - calculated_values[-1]) > \
+                    abs(calculated_values[-2] - calculated_values[-3]):
+                return False, i
+
+            if num_of_calculated_vals >= 2 and \
+                    abs(calculated_values[-2] - calculated_values[-1]) < CONVERGENCE_THRESHOLD:
+                return True, i
+
+    return False, i
+
+
+class FREnumerator(RelativeGCFEnumerator):
+    """
+    This enumerator checks the Factorial Reduction property of GCFs as the first step of the enumeration.
+    In the FR test we don't compute the GCF's value, or even compare it to a LHS.
+    Fractions that have FR will be computed to a higher dept using RelativeGCFEnumerator's implementation.
+    The computed values are then fed into a PSLQ that tries to find a suitable LHS.
+    """
+
+    def __init__(self, *args, **kwargs):
+        print('checking for FR enumerator')
+        super().__init__(None, *args, **kwargs)
+
+    def _first_enumeration(self, print_results: bool):
+        """
+        Test all GCFs in the domain for FR.
+        """
+        results = []  # list of intermediate results        
+        all_items_calculated = []
+        for an_iter, bn_iter, metadata in self._iter_domains_with_cache(FIRST_ENUMERATION_MAX_DEPT):
+            has_fr, items_calculated = check_for_fr(an_iter, bn_iter, self.poly_domains.get_an_degree(metadata.an_coef))
+            if has_fr:
+                all_items_calculated.append(items_calculated)
+                if print_results:
+                    print(f"found a GCF with FR:\n\tan: {metadata.an_coef}\n\tbn: {metadata.bn_coef}")
+                # Key is useless here :)
+                results.append(Match(metadata.an_coef, metadata.bn_coef))
+
+        return results
+
+    def _improve_results_precision(self, intermediate_results, verbose=True):
+        """
+        Calculates GCFs to a higher dept using RelativeGCFEnumerator's implementation.
+        We then feed those results and the constant given to a PSLQ, that tries to find a suitable LHS.
+
+        Notice-
+        The second part of this function (PSLQ), logically belongs to the next step of the algorithm - the 
+        result refinement part. It is implemented here, because the next function is not parallelized over
+        different processes or clients, and we want the PSLQ to be parallelized as well. 
+        """
+        precise_intermediate_results = super()._improve_results_precision(intermediate_results, verbose)
+
+        pslq_results = []
+        consts = [i() for i in self.constants_generator]
+        for match, val, precision in precise_intermediate_results:
+            mpf_val = mpmath.mpf(val)
+            pslq_res = mpmath.pslq(
+                [1, consts[0], consts[0] * consts[0], -mpf_val, -consts[0] * mpf_val, -consts[0] * consts[0] * mpf_val],
```

Comment:
did you mean to use only consts[0]? 
if so, it worth writing something like:

const = self.constants_generator[0]()     # using only 1 constant for now

Reply:
The code tries to find integer coefficients (a1 ... a6) such that for pi: a1 + a2*pi + a3*pi² + tol = a4*pi' + a5*pi'*pi + a6*pi'*pi² for a low tolerance tol. Equivalently: (a1 - a4*pi') + (a2 - a5*pi')*pi + (a3 - a6*pi')*pi² < tol. That is why they are only interested in a single constant (in my example pi).